### ISSUE TITLE
CatBoostClassifier predict() returns type np.float64 irrespective of input type

### ISSUE THREAD

**Issue by vivekk-ezdi** [catboost/catboost#305]
**System Information**
Linux-3.16.0-77-generic-x86_64-with-Ubuntu-14.04-trusty
('Python', '2.7.6 (default, Nov 23 2017, 15:49:48) \n[GCC 4.8.4]')
('NumPy', '1.14.2')
('SciPy', '1.0.1')
('Scikit-Learn', '0.19.1')
('Catboost', '0.7.2.3')

**Description**:

`CatBoostClassifier` `predict()` only returns dtype `np.float64`.

I understand that the according to the [documentation of `fit()` method](https://tech.yandex.com/catboost/doc/dg/concepts/python-reference_catboost_fit-docpage/), the proper input labels should be `float`, so the corresponding output will also be `float`, but the output is not documented.

>
>y
> 
>     Must be in the form of a one-dimensional array of floating point numbers.

But since the [description of `CatBoostClassifier`](https://tech.yandex.com/catboost/doc/dg/concepts/python-reference_catboostclassifier-docpage/) has this:

> 
> Purpose
> 
>     Training and applying models for the classification problems. 
>     When using the applying methods only the probability that the object belongs to the class is returned. 
>     Provides compatibility with the scikit-learn tools.

This presents us with **The input type and output type mismatch** problem:

First, in all scikit classifiers, the labels are encoded to appropriate type internally using [`LabelEncoder`](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html). Secondly, all scikit classifiers return the output type same as input types (due to internal encoding as descibed above).

```
import numpy as np
import pandas as pd

from sklearn.datasets import make_classification
X, y = make_classification()
X = np.abs(X)

# Utility dataframe to store different types of labels
y_all = pd.DataFrame(np.zeros((y.shape[0],4)), columns=['int', 'bool', 'str', 'float'])
y_all['int'] = y
y_all['bool'] = y>0
y_all['str'] = np.asarray(['a' if yy>0 else 'b' for yy in y], dtype=str)
y_all['float'] = np.asarray(y, dtype=np.float64) 
```

**Example demonstrating the current working of all scikit estimators**:
```
from sklearn.base import ClassifierMixin
from sklearn.utils.testing import all_estimators
classifiers=[est[1]() for est in all_estimators() if issubclass(est[1], ClassifierMixin)]
for clf in classifiers:
    for i in ['int', 'bool', 'str', 'float']:
        clf.fit(X, y_all[i].values)
        pred = clf.predict(X)
        print(type(clf).__name__,y_all[i].values.dtype, pred.dtype)

Output:
('AdaBoostClassifier', dtype('int64'), dtype('int64'))
('AdaBoostClassifier', dtype('bool'), dtype('bool'))
('AdaBoostClassifier', dtype('O'), dtype('O'))
('AdaBoostClassifier', dtype('float64'), dtype('float64'))
('BaggingClassifier', dtype('int64'), dtype('int64'))
('BaggingClassifier', dtype('bool'), dtype('bool'))
('BaggingClassifier', dtype('O'), dtype('O'))
('BaggingClassifier', dtype('float64'), dtype('float64'))
('BernoulliNB', dtype('int64'), dtype('int64'))
...
...
...
```

Now `CatBoostClassifier` accepts labels of type **(`int, float, bool`)** and we need to explicitly convert the other type labels (strings), but documentation mentions **only `float` input**.
Then `CatBoostClassifier` returns only `float` irrespective of input type from (`int, float, bool`).

**Example demonstrating the input and output type in CatBoostClassifier**:
```
from catboost import CatBoostClassifier
clf = CatBoostClassifier(verbose=False)
for i in ['int', 'bool', 'str', 'float']:
    try:
        clf.fit(X, y_all[i].values)
        pred = clf.predict(X)
        print(pred.shape)
        print(y_all[i].values.dtype, pred.dtype)
    except Exception as e:
        print(e)

Output:
(dtype('int64'), dtype('float64'))
(dtype('bool'), dtype('float64'))
could not convert string to float: a
(dtype('float64'), dtype('float64'))
```

This make the usage of `CatBoostClassifier` difficult when used in a Pipeline or ensembles like VotingClassifier. Even when the input type is `float`.

**Example demonstrating the error in VotingClassifier due to float output**:
```
from sklearn.ensemble import RandomForestClassifier, VotingClassifier
clf1 = CatBoostClassifier(verbose=False)
clf2 = RandomForestClassifier()
clf = VotingClassifier(estimators=[('cb', clf1), ('rf', clf2)])
for i in ['int', 'bool', 'str', 'float']:
    try:
        clf.fit(X, y_all[i].values)
        pred = clf.predict(X)
        print(pred.shape)
        print(y_all[i].values.dtype, pred.dtype)
    except Exception as e:
        print(e)

Output:
Cannot cast array data from dtype('float64') to dtype('int64') according to the rule 'safe'
Cannot cast array data from dtype('float64') to dtype('int64') according to the rule 'safe'
Cannot cast array data from dtype('float64') to dtype('int64') according to the rule 'safe'
Cannot cast array data from dtype('float64') to dtype('int64') according to the rule 'safe'
```

**Proposed Solution**
Recommended solution would be that the `CatBoostClassifier` should handle the encoding and decoding of labels internally same as other scikit classifiers.  Something like:

```
class CatBoostClassifierCorrected(CatBoostClassifier):
    def fit(self, X, y=None, cat_features=None, sample_weight=None, baseline=None, use_best_model=None,
        eval_set=None, verbose=None, logging_level=None, plot=False, column_description=None, verbose_eval=None):
        # Handle different types of label
        self.le_ = LabelEncoder().fit(y)
        transformed_y = self.le_.transform(y)

        self._fit(X, transformed_y, cat_features, None, sample_weight, None, None, None, baseline, use_best_model, eval_set, verbose, logging_level, plot, column_description, verbose_eval)
        return self
        
    def predict(self, data, prediction_type='Class', ntree_start=0, ntree_end=0, thread_count=1, verbose=None):
        predictions = self._predict(data, prediction_type, ntree_start, ntree_end, thread_count, verbose)

        # Return same type as input
        return self.le_.inverse_transform(predictions.astype(np.int64))
```

But I understand that it will be a long process with testing etc, so the current proposal is that the documentation should be made clear as to what goes in and what comes out. Also, we can throw a warning when the input type is cast internally from `int` and  `bool` to `float`.

**Comment by annaveronika** [comment]
Adding help wanted flag, please contribute if this feature is important for you!

**Comment by andrey-khropov** [comment]
It has been fixed by commit https://github.com/catboost/catboost/commit/f2d4cc2aa3b9a6421b731dd97634ec2c3191c181, the release with this fix included is expected within a day.
